core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -74,134 +35,346 @@ class Queue[A] private (
+      def offerAll(as: Iterable[A]): IO[E2, Boolean] = self.offerAll(as)
+      def awaitShutdown: IO[E2, Unit]                = self.awaitShutdown
+      def size: IO[E2, Int]                          = self.size
+      def shutdown: IO[E2, Unit]                     = self.shutdown
+      def take: IO[E2, C]                            = self.take.flatMap(f)
+      def takeAll: IO[E2, List[C]]                   = self.takeAll.flatMap(IO.foreach(_)(f))
+      def takeUpTo(max: Int): IO[E2, List[C]]        = self.takeUpTo(max).flatMap(IO.foreach(_)(f))
+    }
+
+  def zipWith[E2 >: E, A2 <: A, C, D](that: Queue2[E2, A2, C])(f: (B, C) => D): Queue2[E2, A2, D] =
+    new Queue2[E2, A2, D] {
+      def capacity: Int = self.capacity
+
+      // TODO: self.offer orElse that.offer? Do we need tryOffer?
+      def offer(a: A2): IO[E2, Boolean] = self.offer(a)
+      // TODO: Offer everything or split?
+      def offerAll(as: Iterable[A2]): IO[E2, Boolean] = self.offerAll(as)
+
+      def awaitShutdown: IO[E2, Unit]                 = self.awaitShutdown
+      def size: IO[E2, Int]                           = self.size
+      def shutdown: IO[E2, Unit]                      = self.shutdown
+      def take: IO[E2, D]                             = self.take.zipWith(that.take)(f)
+
+      def takeAll: IO[E2, List[D]] =
+        self.takeAll.zipWith(that.takeAll) { (bs, cs) =>
+          val bsIt = bs.iterator
+          val csIt = cs.iterator
+
+          bsIt.zip(csIt).map(f.tupled).toList
+        }
+
+      def takeUpTo(max: Int): IO[E2, List[D]] =
+        self.takeUpTo(max).zipWith(that.takeUpTo(max)) { (bs, cs) =>
+          val bsIt = bs.iterator
+          val csIt = cs.iterator
+
+          bsIt.zip(csIt).map(f.tupled).toList
+        }
+    }
 
-      as <- IO.sync0 { context =>
-             val as = unsafePollN(queue, max)
-             strategy.unsafeOnQueueEmptySpace(queue, context)
-             as
-           }
-    } yield as
+  def zip[E2 >: E, A2 <: A, C](that: Queue2[E2, A2, C]): Queue2[E2, A2, (B, C)] =
+    zipWith(that)((_, _))
+  
+  def filter(f: B => Boolean): Queue2[E, A, B] =
+    new Queue2[E, A, B] {
+      def capacity: Int                             = self.capacity
+      def offer(a: A): IO[E, Boolean]               = self.offer(a)
+      def offerAll(as: Iterable[A]): IO[E, Boolean] = self.offerAll(as)
+      def awaitShutdown: IO[E, Unit]                = self.awaitShutdown
+      def size: IO[E, Int]                          = self.size
+      def shutdown: IO[E, Unit]                     = self.shutdown
+      def take: IO[E, B] = {
+        def retry: IO[E, B] =
+          self.take.flatMap { b =>
+            if (f(b)) IO.succeed(b)
+            else retry
+          }
+
+        retry
+      }
+
+      def takeAll: IO[E, List[B]] =
+        self.takeAll.map(_.filter(f))
+
+      // TODO: rethink semantics here; maybe spin until we've got max or queue is empty
```

Comment:
@jdegoes @ghostdogpr

Reply:
Yeah, I think it makes sense to retry with `max - current valid items` until we have `max` items or `takeUpTo` returns empty.